### PR TITLE
[WIP] Proxy

### PR DESCRIPTION
**Pull request by kelvich**
General description is here: https://github.com/zenithdb/console/blob/main/docs/proxy.md

TODO for this PR: handle socket termination that is commented out right now

### CODE REVIEW

**Review by funbringer — COMMENTED** [review]
Some comments so far. There's more to come.

File: proxy/src/proxy.rs [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+        conn.handle_new_user()?
+    };
+
+    // ok, proxy pass user connection to database_uri
+    let runtime = tokio::runtime::Builder::new_current_thread()
+        .enable_all()
+        .build()
+        .unwrap();
+
+    let _ = runtime.block_on(proxy_pass(conn.pgb, db_info))?;
```
Comment (funbringer) [inline proxy/src/proxy.rs:86]: Apparently, you're fighting a leaking async abstraction.

File: zenith_utils/src/postgres_backend.rs [hunk 2]
```diff
@@ -112,28 +150,80 @@ impl PostgresBackend {
...
+                                    self.write_message_noflush(&BeMessage::AuthenticationOk)?;
+                                    // psycopg2 will not connect if client_encoding is not
+                                    // specified by the server
+                                    self.write_message_noflush(&BeMessage::ParameterStatus)?;
+                                    self.write_message(&BeMessage::ReadyForQuery)?;
+                                    self.state = ProtoState::Established;
+                                }
+                                AuthType::MD5 => {
+                                    rand::thread_rng().fill(&mut self.md5_salt);
+                                    let md5_salt = self.md5_salt.clone();
```
[the remark this answers is not in the document]
Comment (kelvich) [reply zenith_utils/src/postgres_backend.rs]: fixed

File: proxy/src/cplane_api.rs [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+        );
+
+        if received_hash == salted_stored_hash {
+            Ok(())
+        } else {
+            bail!("Auth failed")
+        }
+    }
+
+    pub fn get_database_uri(&self, _user: &String, _database: &String) -> Result<DatabaseInfo> {
```
[the remark this answers is not in the document]
Comment (kelvich) [reply proxy/src/cplane_api.rs]: fixed

File: proxy/src/proxy.rs [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+databases without opening the browser.
+
+", self.state.conf.redirect_uri,self.psql_session_id);
+
+        self.pgb
+            .write_message_noflush(&BeMessage::AuthenticationOk)?;
+        self.pgb
+            .write_message_noflush(&BeMessage::ParameterStatus)?;
+        self.pgb
+            .write_message(&BeMessage::NoticeResponse(hello_message.to_string()))?;
```
[the remark this answers is not in the document]
Comment (kelvich) [reply proxy/src/proxy.rs]: fixed

File: proxy/src/main.rs [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+
+    threads.push(
+        thread::Builder::new()
+            .name("Mgmt thread".into())
+            .spawn(move || mgmt::thread_main(&state, mgmt_listener))?,
+    );
+
+    for t in threads {
+        let _ = t.join().unwrap();
+    }
```
[the remark this answers is not in the document]
Comment (kelvich) [reply proxy/src/main.rs]: fixed

File: proxy/src/main.rs [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+        waiters: Mutex::new(HashMap::new()),
+    };
+    let state: &'static ProxyState = Box::leak(Box::new(state));
+
+    // Check that we can bind to address before further initialization
+    println!("Starting proxy on {}", state.conf.proxy_address);
+    let pageserver_listener = TcpListener::bind(state.conf.proxy_address)?;
+
+    println!("Starting mgmt on {}", state.conf.mgmt_address);
+    let mgmt_listener = TcpListener::bind(state.conf.mgmt_address)?;
```
[the remark this answers is not in the document]
Comment (kelvich) [reply proxy/src/main.rs:88]: That was intentional -- if we check that port isn't occupied before starting actual threads it is easier to shutdown with proper error message. Updated comment to make this clear.

File: proxy/src/mgmt.rs [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+    ) -> anyhow::Result<()> {
+        let mut query_string = query_string.to_vec();
+        if let Some(ch) = query_string.last() {
+            if *ch == 0 {
+                query_string.pop();
+            }
+        }
+
+        println!("Got mgmt query: '{}'", std::str::from_utf8(&query_string)?);
+
```
[the remark this answers is not in the document]
Comment (kelvich) [reply proxy/src/mgmt.rs:89]: fixed

File: proxy/src/proxy.rs [hunk 8]
```diff
@@ -0,0 +1,256 @@
...
+                                .into();
+                            self.database = m
+                                .params
+                                .get("database")
+                                .ok_or_else(|| {
+                                    anyhow::Error::msg("database is required in startup packet")
+                                })?
+                                .into();
+
+                            self.existing_user = self.user.ends_with("@zenith");
```
[the remark this answers is not in the document]
Comment (kelvich) [reply proxy/src/proxy.rs]: fixed

File: proxy/src/proxy.rs (hunk 1 above)
Comment (kelvich) [reply proxy/src/proxy.rs:84]: Not sure I understood what you are talking about.

Frontal part of proxy done via threads now as PostgresBackend was intended for threaded use, but we should switch to all async here, that is in TODO.

**Pull request merged**